### DISCUSSION TITLE
Audio support

### DISCUSSION THREAD

**Discussion by samjabrahams** [foxglove/community#173] [Q&A] [answered]
hello

i am building (really planning on building I just need to make a few more design tweaks first) **very sophisticated** a cappella (VOICES ONLY) robots. i have attached prototype designs for these robots below (PLEASE DO NOT TAKE THIS IDEA)

![acabots](https://user-images.githubusercontent.com/11607205/126591064-3e245a94-7b52-4633-977a-327a574c1772.png)

anyways the robot has audio and right now i dont see a way to play audio from foxglove. it would be nice to hear my acabots sing in foxglove. i didnt see an audio ticket so i made this one.

thx

(Saw your Show HN post! Hope you all are doing well :D)

**Comment by jtbandes** [comment]
Good feature idea and great robot idea—with this feature Foxglove Studio could really sing!!

What datatypes do these hypothetical acabots use?

Possibly relevant ROS docs: http://wiki.ros.org/audio_common

**Comment by samjabrahams** [comment]
so im personally a fan of raw wav files because i LOVE the clarity that pure uncompressed audio that can provide. my ears are pretty good. but mp3s are nice too some people say you can't tell the difference.

audio common is ok but the main annoying thing is that the raw data and the actual metadata about the audio are in two separate message types!! that means you need two topics to stream the data unless you know the codec ahead of time. but it is a bit more streamlined for streaming data in chunks since you don't need the overhead of resending the codec on every message. probably not a bad place to start since it seems to be the main "audio" library of ros. you can also always roll your own message types, but now we're getting COMPLICATED 🤷 🤷

**Comment by BRNKR** [comment]
any news regarding audio ? what about sending audio to the robot? :)

**Comment by jtbandes** [comment]
Hi @TobiMiller, thanks for your interest! No news to share on this feature currently. Can you give more details about specifically what features and data formats you would need?

**Comment by BRNKR** [comment]
@jtbandes Thank you for your answer. For our research project, we need a two-way audio communication to the robot. Here we want a teleoperator to be able to talk to humans around the robot and have them respond as well. So far we have audio_capture and audio_play nodes running on each side and could also communicate via a first attempt to write our own web interface. But meanwhile I think Floxglove would be a much nicer solution for this. If a plugin is available for "audio_common" messages. As I said. Two-way audio communication is what we need.

**Comment by BRNKR** [comment]
@jtbandes does the foxglove dev team have time for such a plugin? Otherwise we would try to build a prototype ourselves. However, none of us in the team masters typescript. so it would take some time. we would appreciate it very much if it is no effort for you to wait for now.

**Comment by amacneil** [comment]
This isn't currently on our roadmap - we usually prioritize features based on number of people asking for them.

If you create your own plugin though, we're happy to offer help/advice! ([join our Slack](https://foxglove.dev/slack) for live support)

**Comment by jaagut** [comment]
We at @bit-bots also use the `audio_common` package, as it is a _common_ standard for audio messages in ROS 2 and would really appreciate support for this. Btw, Foxglove Studio is really awesome software and we are always discovering new features!

Some additional thoughts on what the audio panel could do:

- Playback
    - in current ROSbag-playing speed,
    - multiple (mutable) channels,
    - stereo playback
- displaying the raw wave-form for every channel
- displaying the transformed spectrogram

**Reply by jtbandes** [reply]
Hi @jaagut, would you be able to provide some sample bags/datasets that contain audio data? I don't think we've encountered any real-world data so far.

**Reply by jaagut** [reply]
Sure, all or most of these should contain audio messages: https://data.bit-bots.de/ROSbags/robocup_2023/

Just for reference, as a RoboCup Soccer Team, we are currently working on a whistle sound detection and are therefore interested in easily investigating the audio recordings and correlating it with detections.

**Comment by thomakl** [comment]
Hi everyone,

Do you have any news on the developpement of an audio plugin ?
That would be very useful.

Cheers,

**Reply by defunctzombie** [reply]
We are working on it - do you currently record audio? If so - in what format(s)? How?

**Comment by thomakl** [comment]
Hello,
That's excellent news !
I do record raw audio signal with a third-party software connected to Foxglove.
It renders a stream of raw data of Floats 32.
If you need to discuss more about it, let me know.

I would be happy to help by testing the development of this plugin.

**Reply by defunctzombie** [reply]
We don't yet have the panel ready for testing but the schema we are proposing for recording raw audio is documented here: https://github.com/foxglove/foxglove-sdk/tree/main/schemas#rawaudio

**Reply by ComePerin** [reply]
Hello, I am also very interested in this plugin. Is there a way I could see its status (maybe a repository ?), or even contribute to its development ?

**Comment by JordanBrindzaScout** [comment]
@defunctzombie I am also interested in audio support but from the other direction, record audio through foxglove and send to the robot. We have our extension working and recording in chrome but are running into issues getting microphone access in the app, will the plugin you are working on address this issue? Thank you

discord thread ref: https://discord.com/channels/1240935776649478144/1379561598267424768/1379561598267424768

**Reply by defunctzombie** [reply]
Right now the feature we are working on is for playing audio in the app not for recording audio in the app.

**Comment by amacneil** [comment, marked as answer]
Audio support is live!

https://foxglove.dev/blog/announcing-audio-panel

**Comment by samjabrahams** [comment]
Incredible! What a journey. Since first opening this ticket, I have pivoted my original Acabot concept to an AI powered SaaS (singing as a service), specifically targeting the four-part harmony gap in the market. I've put the bots to work folding laundry. Congrats, Foxglove team!
